redbiom/fetch.py

```diff
@@ -0,0 +1,271 @@
+    import json
+    from operator import itemgetter
+    import scipy.sparse as ss
+    import biom
+    import redbiom.requests
+
+    # TODO: centralize this as it's boilerplate
+    if get is None:
+        import redbiom
+        config = redbiom.get_config()
+        get = redbiom.requests.make_get(config)
+
+    redbiom.requests.valid(context, get)
+
+    # pull out the observation index so the IDs can be remapped
+    obs_index = json.loads(get(context, 'GET', '__observation_index'))
+
+    # redis contains {observation ID -> internal ID}, and we need
+    # {internal ID -> observation ID}
+    inverted_obs_index = {v: k for k, v in obs_index.items()}
+
+    # pull out the per-sample data
+    table_data = []
+    unique_indices = set()
+    getter = redbiom.requests.buffered(samples, 'data', 'MGET', context,
+                                       get=get, buffer_size=100)
+    for (sample_set, sample_set_data) in getter:
+        for sample, data in zip(sample_set, sample_set_data):
+            data = data.split('\t')
+            table_data.append((sample, data))
+
+            # update our perspective of total unique observations
+            unique_indices.update({int(i) for i in data[::2]})
+
+    # construct a mapping of
+    # {observation ID : index position in the BIOM table}
+    unique_indices_map = {observed: index
+                          for index, observed in enumerate(unique_indices)}
+
+    # pull out the observation and sample IDs in the desired ordering
+    obs_ids = [inverted_obs_index[k]
+               for k, _ in sorted(unique_indices_map.items(),
+                                  key=itemgetter(1))]
+    sample_ids = [d[0] for d in table_data]
+
+    # fill in the matrix
+    mat = ss.lil_matrix((len(unique_indices), len(table_data)))
+    for col, (sample, col_data) in enumerate(table_data):
+        # since this isn't dense, hopefully roworder doesn't hose us
+        for index, value in zip(col_data[::2], col_data[1::2]):
+            mat[unique_indices_map[int(index)], col] = value
+
+    return biom.Table(mat, obs_ids, sample_ids)
+
+
+def category_sample_values(category, samples=None):
+    """Obtain the samples and their corresponding category values
+
+    Parameters
+    ----------
+    category : str
+        A metadata column of interest.
+    samples : Iterable of str, optional
+        If provided, only the specified samples and their values are obtained.
+
+    Returns
+    -------
+    pandas.Series
+        A Series indexed by the Sample ID and valued by the metadata value for
+        that sample for the specified category.
+
+    Redis command summary
+    ---------------------
+    HGETALL metadata:category:<category>
+    HMGET metadata:category:<category> <sample_id> ... <sample_id>
+    """
+    import redbiom
+    import redbiom.requests
+    import pandas as pd
+
+    get = redbiom.requests.make_get(redbiom.get_config())
+
+    key = 'category:%s' % category
+    if samples is None:
```

Comment:
The `else` block is triggered, which will return an empty dict resulting in this method returning an empty series

```python
In [4]: redbiom.fetch.category_sample_values('BarcodeSequence', [])
Out[4]: Series([], dtype: float64)
```